network/data_link/at_parser.c

```diff
@@ -0,0 +1,1145 @@
+	callback_desc.callback =
+		(void (*)(void*, uint32_t, void*))at_callback;
+	callback_desc.ctx = ldesc;
+	callback_desc.config = param->uart_irq_conf;
+	if (SUCCESS != irq_register_callback(ldesc->irq_desc,
+					     ldesc->uart_irq_id,
+					     &callback_desc))
+		goto free_desc;
+
+	if (SUCCESS != irq_enable(ldesc->irq_desc, ldesc->uart_irq_id))
+		goto free_irq;
+
+	/* The read will be handled by the callback */
+	uart_read_nonblocking(ldesc->uart_desc, &ldesc->read_ch, 1);
+
+	/* Link buffer structure with static buffers */
+	ldesc->result.buff = ldesc->buffers.result_buff;
+	ldesc->result.len = 0;
+	ldesc->cmd.buff = ldesc->buffers.cmd_buff;
+	ldesc->cmd.len = CMD_BUFF_LEN;
+
+	if (SUCCESS != cb_init(&ldesc->cb_response, RESULT_BUFF_LEN,
+			       sizeof(uint8_t)))
+		goto free_irq;
+
+	ldesc->callback_operation = READING_RESPONSES;
+
+	/* Disable echoing response */
+	if (SUCCESS != stop_echo(ldesc))
+		goto free_cb;
+
+	/* Test AT */
+	if (SUCCESS != at_run_cmd(ldesc, AT_ATTENTION, AT_EXECUTE_OP, NULL))
+		goto free_cb;
+
+	/* Get the connection type */
+	if (SUCCESS == at_run_cmd(ldesc, AT_SET_CONNECTION_TYPE, AT_QUERY_OP,
+				  &result)) {
+		/* ldesc->cmd used as temporary buffer */
+		cb_size(result.out.result, &ldesc->cmd.len);
+		cb_read(result.out.result, ldesc->cmd.buff, ldesc->cmd.len);
+		at_sscanf(&ldesc->cmd, PUI8("+CIPMUX:%u\r\n"), &conn, NULL);
+		ldesc->cmd.len = 0;
+		ldesc->multiple_conections = conn ? true : false;
+	} else {
+		goto free_cb;
+	}
+
+	*desc = ldesc;
+	return SUCCESS;
+
+free_cb:
+	cb_remove(ldesc->cb_response);
+free_irq:
+	irq_unregister(ldesc->irq_desc, ldesc->uart_irq_id);
+free_desc:
+	free(ldesc);
+	*desc = NULL;
+	return FAILURE;
+}
+
+/**
+ * @brief Free resources allocated at \ref at_init
+ * @param desc - AT parser reference
+ * @return
+ *  - \ref SUCCESS : On success
+ *  - \ref FAILURE : Otherwise
+ */
+int32_t at_remove(struct at_desc *desc)
+{
+	if (!desc)
+		return FAILURE;
+	cb_remove(desc->cb_response);
+	irq_unregister(desc->irq_desc, desc->uart_irq_id);
+	free(desc);
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Convert null terminated string to at_buff
+ * @param dest - Destination buffer
+ * @param src - Source buffer
+ * @return
+ *  - \ref SUCCESS : On success
+ *  - \ref FAILURE : Otherwise
+ */
+int32_t str_to_at(struct at_buff *dest, const uint8_t *src)
+{
+	if (!dest || !src)
+		return FAILURE;
+
+	dest->buff = (uint8_t *)src;
+	dest->len = strlen((char *)src);
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Convert at_buff to null terminated string
+ * @param dest - Destination buffer
+ * @param src - Source buffer
+ * @return
+ *  - \ref SUCCESS : On success
+ *  - \ref FAILURE : Otherwise
+ */
+int32_t at_to_str(uint8_t **dest, const struct at_buff *src)
+{
+	if (!src || !dest)
+		return FAILURE;
+
+	src->buff[src->len] = 0;
+	*dest = src->buff;
+
+	return SUCCESS;
+}
+
+/* copy src into dest */
+static inline void at_strcpy(struct at_buff *dest, const struct at_buff *src)
```

Comment:
remove these function and do directly, makes more sense when viewing the code.

Reply:
Removed